### PR TITLE
Validate response variable

### PR DESCRIPTION
**Pull request by maheshakya**
Model building stops when categorical response variable is used with numerical prediction algorithms.
Proper error messages are provided.

### CODE REVIEW

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java [hunk 1]
```diff
@@ -86,6 +87,14 @@ public MLModel build() throws MLModelBuilderException {
             Workflow workflow = context.getFacts();
             long modelId = context.getModelId();
 
+            // Verify validity of response variable
+            // Stops model building if a categorical attribute is used with numerical prediction
+            if (workflow.getAlgorithmClass().equals("Numerical_Prediction") &&
+                    getTypeOfResponseVariable(workflow.getResponseVariable(),
+                            workflow.getFeatures()).equals("CATEGORICAL"))
+                throw new MLModelBuilderException("A categorical attribute cannot be used as the response variable " +
```
Comment (nirmal070125) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java]: Can we be specific on the error? What algorithm? which response variable? etc.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java [hunk 2]
```diff
@@ -116,46 +125,46 @@ public MLModel build() throws MLModelBuilderException {
...
+                case RIDGE_REGRESSION:
+                    summaryModel = buildRidgeRegressionModel(sparkContext, modelId, trainingData, testingData, workflow,
+                            mlModel, includedFeatures);
+                    break;
+                case LASSO_REGRESSION:
+                    summaryModel = buildLassoRegressionModel(sparkContext, modelId, trainingData, testingData, workflow,
+                            mlModel, includedFeatures);
+                    break;
+                default:
+                    throw new AlgorithmNameException("Incorrect algorithm name");
```
Comment (nirmal070125) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java]: Please don't format irrelevant code with a different purpose commit, as it clutters your real fix and makes it hard to review. Are you using WSO2 code template?

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java [hunk 3]
```diff
@@ -167,6 +176,15 @@ public MLModel build() throws MLModelBuilderException {
         }
     }
 
+    private String getTypeOfResponseVariable(String responseVariable, List<Feature> features){
+        String type = null;
+        for(Feature feature: features){
+            if(feature.getName().equals(responseVariable))
+                type = feature.getType();
+        }
+        return type;
```
Comment (nirmal070125) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java:194]: This method could return null right? Then please handle it properly in the place it is used.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java [hunk 4]
```diff
@@ -708,7 +733,7 @@ private MulticlassConfusionMatrix getMulticlassConfusionMatrix(MulticlassMetrics
      * @param predictionsAndLabels Prediction and label values RDD
      */
     private RegressionMetrics getRegressionMetrics(JavaSparkContext sparkContext,
-            JavaRDD<Tuple2<Double, Double>> predictionsAndLabels) {
+                                                   JavaRDD<Tuple2<Double, Double>> predictionsAndLabels) {
```
Comment (nirmal070125) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java]: Better to send a separate PR for formatting purposes, if you are using WSO2 Code template.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java [hunk 5]
```diff
@@ -86,6 +87,21 @@ public MLModel build() throws MLModelBuilderException {
             Workflow workflow = context.getFacts();
             long modelId = context.getModelId();
 
+            // Verify validity of response variable
+            String typeOfResponseVariable = getTypeOfResponseVariable(workflow.getResponseVariable(),
+                    workflow.getFeatures());
+
+            if(typeOfResponseVariable == null && workflow.getAlgorithmClass() != "Clustering")
```
Comment (nirmal070125) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java]: Instead of hard coding Clustering, can you please use AlgorithmType enum.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java [hunk 6]
```diff
@@ -86,6 +89,21 @@ public MLModel build() throws MLModelBuilderException {
             Workflow workflow = context.getFacts();
             long modelId = context.getModelId();
 
+            // Verify validity of response variable
+            String typeOfResponseVariable = getTypeOfResponseVariable(workflow.getResponseVariable(),
+                    workflow.getFeatures());
+
+            if(typeOfResponseVariable == null)
```
Comment (nirmal070125) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java:96]: Braces are missing.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java [hunk 7]
```diff
@@ -86,6 +89,21 @@ public MLModel build() throws MLModelBuilderException {
...
+                throw new MLModelBuilderException("Type of response variable cannot be null for supervised learning" +
+                        "algorithms.");
+
+            // Stops model building if a categorical attribute is used with numerical prediction
+            if (workflow.getAlgorithmClass().equals(AlgorithmType.NUMERICAL_PREDICTION) &&
+                    typeOfResponseVariable.equals(FeatureType.CATEGORICAL))
```
Comment (nirmal070125) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java:102]: Braces are missing.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java [hunk 8]
```diff
@@ -167,6 +185,15 @@ public MLModel build() throws MLModelBuilderException {
         }
     }
 
+    private String getTypeOfResponseVariable(String responseVariable, List<Feature> features){
+        String type = null;
+        for(Feature feature: features){
+            if(feature.getName().equals(responseVariable))
```
Comment (nirmal070125) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java:191]: Braces are missing.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java (hunk 3 above)
Comment (nirmal070125) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java:194]: Did we handle?

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java (hunk 7 above)
Comment (maheshakya) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java:102]: Does not exist in the latest code.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java (hunk 6 above)
Comment (maheshakya) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java:96]: Does not exist in the latest code.

File: components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java (hunk 8 above)
Comment (maheshakya) [inline components/ml/org.wso2.carbon.ml.core/src/main/java/org/wso2/carbon/ml/core/spark/algorithms/SupervisedSparkModelBuilder.java:191]: Fixed in the PR: https://github.com/wso2/carbon-ml/pull/133
